src/editor.js

```diff
@@ -459,6 +544,72 @@ class Editor {
     }
     return str;
   }
+
+  /**
+   * Replaces CodeMirror hints with the AMP spec.
+   *
+   * Kinda stolen from @ampproject/docs/playground/src/editor.js
+   * (Editor#loadHints).
+   *
+   * Additionally sets uploaded file name hints for attrs like "src".
+   *
+   * @param {string} format
+   *    Should always be "amp4ads" for our use case, see top-level keys of
+   *    amphtml-hint.json for possible values.
+   * @private
+   */
+  async setHints_(format) {
+    const hints = await this.amphtmlHints_;
+    for (const key of Object.keys(codemirror.htmlSchema)) {
+      delete codemirror.htmlSchema[key];
+    }
+    Object.assign(codemirror.htmlSchema, hints[format.toLowerCase()]);
+
+    // Below, ours, in case of race:
+    this.updateFileHints_();
+  }
+
+  /**
+   * Loads a JSON file containing the CodeMirror HTML Schema to be consumed for
+   * AMP formats.
+   *
+   * Kinda stolen from @ampproject/docs/playground/src/editor.js
+   * (Editor#fetchHintsData).
+   * @private
+   */
+  fetchHintsData_() {
+    const {promise, reject, resolve} = new Deferred();
+    this.win.requestIdleCallback(async () => {
+      try {
+        const response = await fetch(hintsUrl);
+        assert(response.status === 200, `fetch got ${response.status}`);
+        resolve(response.json());
+      } catch (err) {
+        reject(err);
+      }
+    });
+    return promise;
+  }
+
+  /**
+   * Sets attr hints like for "src" and "poster" for the tags in the spec that
+   * allow it, so that they show uploaded files.
+   * @private
+   */
+  updateFileHints_() {
+    const readableUrls = this.state_.files.map(({name}) => `/${name}`);
+
+    for (const tagName of attrFileHintTagNames) {
+      if (!(tagName in codemirror.htmlSchema)) {
+        // spec not populated yet, will exec again when fetched.
+        return;
+      }
+      const {attrs} = codemirror.htmlSchema[tagName];
+      for (const attr of attrFileHintTagAttrs[tagName]) {
+        attrs[attr] = readableUrls.length > 0 ? [...readableUrls] : null;
```

Comment:
why cloning readableUrls?

Reply:
Mainly for sanity. Technically, `CodeMirror` doesn't expect any of the attribute sets in the spec to have many-to-one references. But it seems to work fine if not cloning, so I'll just do that.